pkg/controller/ipam/ipam_test.go

```diff
@@ -0,0 +1,421 @@
+package ipam
+
+import (
+	"context"
+	"net"
+	"time"
+
+	v1 "github.com/mneverov/cluster-cidr-controller/pkg/apis/clustercidr/v1"
+	clustercidrclient "github.com/mneverov/cluster-cidr-controller/pkg/client/clientset/versioned/typed/clustercidr/v1"
+	clustercidrinformers "github.com/mneverov/cluster-cidr-controller/pkg/client/informers/externalversions"
+	clustercidrinformersv1 "github.com/mneverov/cluster-cidr-controller/pkg/client/informers/externalversions/clustercidr/v1"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeinformers "k8s.io/client-go/informers"
+	informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/komega"
+)
+
+const (
+	timeout  = 2 * time.Second
+	interval = 100 * time.Millisecond
+	resync   = 1 * time.Hour
+)
+
+var _ = ginkgo.Describe("Pod CIDRs", ginkgo.Ordered, func() {
+	var (
+		nodesToCleanup        []*corev1.Node
+		clusterCIDRsToCleanup []*v1.ClusterCIDR
+	)
+	ginkgo.BeforeAll(func() {
+		ctx, cancel = context.WithTimeout(context.Background(), 42*time.Second)
+
+		gomega.SetDefaultConsistentlyDuration(timeout)
+		gomega.SetDefaultConsistentlyPollingInterval(interval)
+		gomega.SetDefaultEventuallyTimeout(timeout)
+		gomega.SetDefaultEventuallyPollingInterval(interval)
+
+		kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, resync)
+		sharedInformerFactory := clustercidrinformers.NewSharedInformerFactory(cidrClient, resync)
+
+		ipamController := booststrapMultiCIDRRangeAllocator(
+			ctx,
+			kubeClient,
+			cidrClient.NetworkingV1().ClusterCIDRs(),
+			kubeInformerFactory.Core().V1().Nodes(),
+			sharedInformerFactory.Networking().V1().ClusterCIDRs(),
+		)
+
+		go ipamController.Run(ctx)
+		sharedInformerFactory.Start(ctx.Done())
+		kubeInformerFactory.Start(ctx.Done())
+
+		ginkgo.DeferCleanup(ginkgo.GinkgoRecover)
+		komega.SetClient(k8sClient)
+		komega.SetContext(ctx)
+	})
+	ginkgo.AfterAll(func() { cancel() })
+
+	ginkgo.BeforeEach(func() {
+		nodesToCleanup = nil
+		clusterCIDRsToCleanup = nil
+	})
+	ginkgo.AfterEach(func() {
+		if len(nodesToCleanup) > 0 {
+			for _, n := range nodesToCleanup {
+				gomega.Expect(
+					kubeClient.CoreV1().Nodes().Delete(ctx, n.Name, metav1.DeleteOptions{}),
+				).To(gomega.Succeed())
+				gomega.Eventually(komega.Get(n)).Should(gomega.WithTransform(apierrors.IsNotFound, gomega.BeTrue()))
+			}
+		}
+		if len(clusterCIDRsToCleanup) > 0 {
```

Comment:
[TestEnv stops](https://github.com/mneverov/cluster-cidr-controller/pull/26/files#diff-57a47481576a21aa5c76d64663cdb224445a678c33e25b53db846ca506ca684eR72) after all tests are finished. So either all nodes and clusterCIDRs should have distinctive names or they should be deleted after each test. Having random names would help, but also I'd prefer to have proper names in logs for debugging.
Common approach (in ginkgo) is to have a [variable](https://github.com/mneverov/cluster-cidr-controller/pull/26/files#diff-3586d4acb1dc04fbe8d5a650fc6a319c60c7e1fd95e741538d82a92fd2f27153R34-R35) declared at the beginning and reusing it in all tests. I used slices since some tests do not create ClusterCIDRs or create one or many Nodes. It is awkward, I agree, but I prefer the cleanup to happen in one place - `AfterEach` instead of cleaning up in each test.
It can be somewhat mitigated if we start a testEnv in each test, but even then [naming](https://github.com/mneverov/cluster-cidr-controller/pull/26/files#diff-3586d4acb1dc04fbe8d5a650fc6a319c60c7e1fd95e741538d82a92fd2f27153R283-R284) should be changed.
Are there other way to do the cleanup?

Reply:
I think deleting all of them after each test makes sense.

May be we can use `cidrClient.NetworkingV1().ClusterCIDRs().List()` in `AfterEach` to list and then delete all the ClusterCIDRs present

Reply:
Agree, thx for the suggestion.